### ISSUE TITLE
fix(SRA download): rule get layout often fails

### ISSUE THREAD

**Issue by uniqueg** [zavolanlab/zarp#166]
**Describe the bug**

The rule `get_layout` in the SRA download workflow often fails. From nine more or less randomly selected libraries, in four cases, no output file (`SINGLE.info` or `PAIRED.info`) is generated. This led to very unpredictable behaviors, in some cases causing the download of libraries to fail for which the library layout was actually successfully determined.

There are at least two issues with the current implementation:
1. The `efetch` command fails to extract the library layout in many cases
2. Failure to fetch the layout does not cause the rule to break and leads to unexpected behaviors

**To Reproduce**

#### From within ZARP-cli

- Install ZARP-cli as per instructions: https://zavolanlab.github.io/zarp-cli/guides/installation/
- Activate the enviroment with `conda activate zarp-cli`
- Execute `zarp SRR23590181` => WORKS
- Execute `zarp SRR23529108` => FAILS
- Execute `zarp SRR23590181 SRR23529108` => NOW _BOTH_ FAIL!
- Execute `zarp SRR21711080` => FAILS
- Execute `zarp SRR23590181 SRR21711080` => NOW FIRST LIB FAILS, BUT SECOND WORKS

#### Isolating the `efetch` problem

- Create Conda environment from `workflow/envs/entrez-direct.yaml`
- Activate environment
- Execute `efetch -db sra -id SRR23590181 -format runinfo | csv2xml -set Set -rec Rec -header | xtract -pattern Rec -if Run -equals SRR23590181 -element LibraryLayout` => WORKS
- Execute `efetch -db sra -id SRR23529108 -format runinfo` => WORKS
- Execute `efetch -db sra -format runinfo -id SRR23529108 | csv2xml -set Set -rec Rec -header` => FAILS
- Execute `efetch -db sra -id SRR23529108 -format runinfo | csv2xml -set Set -rec Rec -header | xtract -pattern Rec -if Run -equals SRR23590181 -element LibraryLayout` => FAILS

In both cases, the error message is:

```console
Mismatched columns in row 2 - 'SRR23529108,2023-02-20 00:39:06,2023-02-19 23:17:49,31254111,9376233300,31254111,300,2782,,https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos4/sra-pub-zq-3/SRR023/23529/SRR23529108/SRR23529108.lite.1,SRX19421567,GSM7051046,RNA-Seq,cDNA,TRANSCRIPTOMIC,PAIRED,0,0,ILLUMINA,HiSeq X Ten,SRP423408,PRJNA936603,,936603,SRS16815738,SAMN33359080,simple,7227,Drosophila melanogaster,GSM7051046,,,,,,,no,,,,,"DR. YONG-ZHEN XU LAB, COLLEGE OF LIFE SCIENCES, WUHAN UNIVERSITY",SRA1592728,,public,ED3F285FD8A10111D732D12FF7818441,D6D719AB91DFAF5455A9C5FD0350A465'
```

This indicates that the CSV to XML conversion fails in many cases. It is not obvious to me why.

**Expected behavior**

The library layout should be correctly extracted in the vast majority of cases. In rare cases, if it cannot be extracted, the rule `get_layout` should fail.

**Screenshots**

N/A

**Additional context**

N/A

**Comment by mkatsanto** [comment]
This issue is caused by the fact that in some fields there are extra commas within quotes. The csv2xml somehow does not recognise that the commas are with the quote and treats them as separators.
An example, where
 efetch -db sra -id SRR23529108 -format runinfo | awk -F'"' -v OFS='' '{ for (i=2; i<=NF; i+=2) gsub(",", "", $i) } 1'  | csv2xml -set Set -rec Rec -header | xtract -pattern Rec -if Run -equals SRR23590181 -element LibraryLayout -> WORKS 

Looking at the https://manpages.debian.org/testing/ncbi-entrez-direct/transmute.1.en.html#encodeXML on which csv2xml is based on , I do not see any direct option that would ignore the commas within a quote. 

Looking at some alternative way of writing this, because the way this rule is now written seems a bit unstable (too many custom specifications).

**Comment by uniqueg** [comment]
Thanks. I thought that additional commas in quoted text might be the issue. And yes, addressing that with `awk` as you suggest would probably help. But I agree with you, it is still going to be messy and might fail with the next unexpected issue. The same of course also goes for the current temporary solution.

In principle though, it's the whole idea of first getting the library layout that's a bit shaky. Because whether a library is `SINGLE` or `PAIRED` is user-submitted - but whether `fastq-dump` produces one, two or more files is depending on the data. These info can _and will_ differ, in some cases. Someone working with HTSinfer told me that there are some SRRs that turn into three files, for example. Plus, indeed there are `PAIRED` samples with just one FASTQ and `SINGLE` samples with more than one.

In other words, we can't fully fix this to account for all cases.

My suggestion: Let's leave the `efetch` implementation for now, until we got some cases where it actually breaks (maybe they are really very rare and nobody will _ever_ complain).

Instead, we can focus on changing the workflow such that it does not fail when there's an issue with individual libraries. Rather, emit some warnings, keep on going silently and just don't update the local paths in the final output. It's easy enough to parse for downstream tools. For example, in ZARP-cli it will be great to have this, because then individual samples are simply dropped if they are, e.g., misspelled. I think that's preferable over breaking the entire workflow (and therefore ZARP-cli).

**Comment by mkatsanto** [comment]
Agree with all suggestions.
One comment on the structure and why I incorporated the get_layout option, for future reference.
Initially I wrote the workflow in a way that I could dynamically (by using a checkpoint) infer whether there is one. or two files and then execute the appropriate workflow. The issue that this caused is that we could not remove the fastq file with a temp flag and expect the workflow to work properly, as each time it needed to recalculate the file would not be there and therefore the whole workflow would start again (In short you cannot use temp flag in a checkpoint rule, reasonably so). This then defeated the whole purpose of compressing the files.

**Comment by mkatsanto** [comment]
In the main workflow, we know from our sample table whether a sample is ".pe" or ".se". So in the final rule we know what to call the files. In this case we learn that within the workflow, so we have to use the checkpoint to 'learn' that.

**Comment by mkatsanto** [comment]
Would it be acceptable if in the dynamic rule we only return these samples for which we could fetch the sample AND found either paired or single in the comments of the library layout? I was never checking if all of the samples have an .info file and thus the instability
